parlai/agents/transformer/module_extensions.py

```diff
@@ -0,0 +1,195 @@
+            default=False,
+            help='Add noise to decoder. At least one of `r3f-encoder-noise` and `r3f-coder-noise` must be set to True',
+        )
+
+    def __init__(self, opts):
+        self.use_r3f = opts.get('use_r3f')
+        self.eps = opts.get('r3f_eps')
+        self.noise_type = opts.get('r3f_noise_type')
+        if self.noise_type == R3F_NOISE_NORMAL:
+            self.noise_sampler = torch.distributions.normal.Normal(
+                loc=0.0, scale=self.eps
+            )
+        elif self.noise_type == R3F_NOISE_UNIFORM:
+            self.noise_sampler = torch.distributions.uniform.Uniform(
+                low=-self.eps, high=self.eps
+            )
+        self.r3f_lambda = opts.get("r3f_lambda")
+        self.noise_encoder = opts.get("r3f_encoder_noise")
+        self.noise_decoder = opts.get("r3f_decoder_noise")
+
+        if self.use_r3f and not self.noise_encoder and not self.noise_decoder:
+            raise RuntimeError(
+                "R3FContext: Noise must be added to at least one of the encoder or decoder."
+            )
+        self.had_first_pass = (
+            False
+        )  # necessary because nn.Embedding for encoder/decoder on some models are the same. Use this to only make any relevant deep copies once.
+
+    def _calculate_symm_kl(self, noised_logits, input_logits):
+        return (
+            F.kl_div(
+                F.log_softmax(noised_logits, dim=-1, dtype=torch.float32),
+                F.softmax(input_logits, dim=-1, dtype=torch.float32),
+                None,
+                None,
+                "sum",
+            )
+            + F.kl_div(
+                F.log_softmax(input_logits, dim=-1, dtype=torch.float32),
+                F.softmax(noised_logits, dim=-1, dtype=torch.float32),
+                None,
+                None,
+                "sum",
+            )
+        ) / noised_logits.size(0)
+
+    def forward_pass_with_r3f(self, model, *args, **kwargs):
+        """
+        'model' argument here is the `nn.Module` normally responsible for the full
+        forward pass.
+        """
+        if not self.use_r3f:
+            raise RuntimeError(
+                "Trying to use an R3F loss when this has not been allowed."
+            )
+        regular_output = model(*args, **kwargs)
+        with R3FNoiseEmbeddingContextManager(self, model) as model:
+            noised_output = model(*args, **kwargs)
+        symm_kl = self._calculate_symm_kl(noised_output[0], regular_output[0])
+        return regular_output, self.r3f_lambda * symm_kl
+
+
+class R3FNoisedEmbedding(nn.Module):
+    def __init__(self, noise_sampler):
+        super(R3FNoisedEmbedding, self).__init__()
+        self.noise_sampler = noise_sampler
+
+    def __call__(self, module, input, output):
+        noise = self.noise_sampler.sample(sample_shape=output.shape).to(output)
+        return output + noise
+
+
+class R3FNoiseEmbeddingContextManager(AbstractContextManager):
```

Comment:
Dope, I think we can merge all 3 of these classes into one now.

Reply:
I think these 3 still need to be separate. Namely, 

1. `R3FNoisedEmbedding` implements `__call__` which is necessary for the forward hook that needs to be separate from the ContextManager to do it (Unless there's some lambda magic that I haven't figured out the incantation for that can be used there)
2. The `AbstractContextManager` needs to be separate from whatever has `forward_pass_with_r3f` (unless having an object use itself in a `with` works?)
3. 

I guess one thing that might make this more clear is if I underscored `R3FNoisedEmbedding` + `R3FNoiseEmbeddingContextManager` and moved them as inner classes to `R3FNoiseContext`? (And change the naming, cause it's kind of confusing as is.)

Reply:
Nvm, just saw your other comment about mixins.

Reply:
The hook can be a member of the context manager like this. `self.` does this automatic currying in python.

```python
class MyContextManger(object):
    def __init__(self, module):
         self.module = module
    def _hook_implementation(self, input, output):
        return output + noise
    def __enter__(self):
        self.module.register_hook(self._hook_implementation)
  ```

Reply:
(For this to combine with the other discussion thread, I think `__enter__` will ultimately have to `return self` to get that access to context manager object)